### ISSUE TITLE
Option to run multi-threaded as opposed to forked?

### ISSUE THREAD

**Issue by mattventura** [prashant-ramcharan/courgette-jvm#75]
Hi,
This plugin does almost exactly what I need, except for the fact that it launches a separate JVM per feature instead of just a thread. On top of using significantly more memory than needed, I need everything to run in a single JVM, as I need to ensure that very specific operations are not performed by multiple threads in parallel (everything else is fine to parallelize). I'm currently doing so with JUnit and maven-surefire, but that has the disadvantage of not being able to parallelize within one runner. Are there any plans to support such a feature? If nothing else, it would be more resource-efficient than spinning up multiple JVMs.

**Comment by prashant-ramcharan** [comment]
Hello,

Courgette is currently designed to address concurrency issues namely thread interference errors and memory consistency errors. 

I felt that it was best to **not** share state between threads and allow each feature / scenario to run in it's own "sandboxed" instance (JVM) thereby removing the need to manage object and class level locks and possible memory leaks from the test code.

Courgette uses the Cucumber CLI which creates a new JVM (_Cucumber world_) for each feature / scenario.  If we use a single JVM, there are race conditions to load Cucumber features, build the Cucumber runtime etc. which throws errors when using a single runner. This is the main reason why a single JVM won't work!

To address the JVM memory issue, you can limit the allocated memory by passing the -Xms and -Xms options to the Courgette runner. [Refer to usage here](https://github.com/prashant-ramcharan/courgette-jvm#gradle-run-options)

Unfortunately, there's no plans to support this feature.

**Comment by mattventura** [comment]
Thanks for the reply. Cucumber currently does support multithreaded testing in a single JVM (via maven-surefire), but it has the downside of not being able to parallel use within a single runner (I.e. you’d need to split the scenarios across at least one runner per thread), meaning you either have to set them up manually, or use a separate plugin to create runners on the fly. 

If you don’t have any plans to support this, I understand.

**Comment by prashant-ramcharan** [comment]
Hi @mattventura 

I used the maven-surefire plugin previously and this was one of the reasons why I created Courgette so that we can have a single runner. 

Another Courgette advantage is that we can parallelise the run on a scenario level with a single runner.

I'll revisit this and see if anything can be done (_cleanly_) to support this feature but as it stands I don't have plans to support it.

**Comment by rhapsodyman** [comment]
When using maven surefire to run several threads in one process, how does it deal with these race conditions ?

**Comment by prashant-ramcharan** [comment]
@rhapsodyman Surefire will only need / create 1 thread which will run Courgette. 

Courgette then creates and manages the forked threads from within the library -- not surefire.

Hope that answers your question.

**Comment by rhapsodyman** [comment]
I was referring to the approach when we have multiple cucumber runner classes - and we execute them in parallel using surefire plugin. I know it is possible to execute them in parallel (one process and multiple threads).

Also wanted to ask - is there any way for courgette to stop its child processes - when we stop the main process ? So in case of long running ui tests - we can stop the execution immediately.
Thank you.

**Comment by prashant-ramcharan** [comment]
In short, the answer is no.  If child processes are executing, it will wait to finish.

Courgette uses an ExecutorService so it cannot guarantee that all the running threads will be stopped at the same time.

When the JVM is killed the ExecutorService will **stop processing new threads** but it will stay alive until all currently executing threads are complete.

**Comment by rhapsodyman** [comment]
Is there any way to add a JVM shutdown hook and stop child processes ? When we stop the courgette from IDE or Jenkins build  - I expect it to stop running cucumber tests. For example - i have 50 features to run: 5 got executed - and i decide to stop the execution - it will not be good to have the remaining features still running after we stop courgette process.

**Comment by prashant-ramcharan** [comment]
If you have 50 features and 5 are being run then decide to stop the execution, only those 5 features will continue to run, the **remaining 45 features will not run.**

The design is, whenever you stop the Courgette execution, only tests are that currently being executed will continue to run until it's complete (_this is due to the ExecutorService_). 

**No new tests** will be started once the Courgette process is stopped.

**Comment by rhapsodyman** [comment]
Got it, thank you.

**Comment by mattventura** [comment]
I ended up basically implementing my own parallelization (with threading rather than forking). Here's some observations:
* It's mostly pointless to run a custom solution for this now, because Cucumber 4+ can natively parallelize within a single runner (Feature level for JUnit, Scenario level for TestNG).
* I implemented a custom solution before that by simply doing the same thing that running Cucumber's main class (the one the plain cucumber CLI would use), which will just set up a separate Cucumber instance each time it's called.
* If you do legitimately need forked execution (very rare but can be useful in some specific scenarios) and want the forks to reliably stop when the main process dies, a shutdown hook isn't the way to go since it only works if the JVM shuts down cleanly. The proper way to do it is by establishing a socket between the main and forked JVMs, and having the forked JVMs exit if the socket is cut.
